### PR TITLE
Socket API

### PR DESCRIPTION
**Pull request by diescalo**
Added support for the new Socket frames and an API to manage them.

### CODE REVIEW

File: digi/xbee/packets/socket.py [hunk 1]
```diff
@@ -0,0 +1,2932 @@
...
+
+    def __set_dest_address(self, dest_address):
+        """
+        Sets the destination address.
+
+        Args:
+            dest_address (Bytearray or String): the new destination address.
+
+        Raises:
+            ValueError: if ``dest_address`` is ``None`` or does not follow the format specified in the configured type.
```
Comment (hgonzaleDigi) [inline digi/xbee/packets/socket.py]: How are exceptions supposed to be described? I have seen that in some occasions, we document different causes of the same exception separately, but other times, like this one, thetwo possible causes of the exception are documented together.
How should this be approached? in any case, either this is the correct form or not, all exceptions should be documented in the same manner.

File: digi/xbee/packets/socket.py [hunk 2]
```diff
@@ -0,0 +1,2932 @@
...
+        Override method.
+
+        .. seealso::
+           | :meth:`.XBeeAPIPacket._get_api_packet_spec_data_dict`
+        """
+        return {DictKeys.SOCKET_ID.value: "%s" % utils.hex_to_string(bytearray([self.__socket_id]))}
+
+    def __get_socket_id(self):
+        """
+        Returns the the socket ID.
```
Comment (hgonzaleDigi) [inline digi/xbee/packets/socket.py]: NP: remove extra "the"

File: digi/xbee/xsocket.py [hunk 3]
```diff
@@ -0,0 +1,814 @@
...
+
+        # If the socket is not created, create it first.
+        if self.__socket_id is None:
+            self.__create_socket()
+
+        lock = threading.Condition()
+        received_state = list()
+
+        # Define the socket state received callback.
+        def socket_state_received_callback(socket_id, state):
```
Comment (hgonzaleDigi) [inline digi/xbee/xsocket.py]: Question, a method defined inside another method is only accessible from the parent method, right? This callback is not actually public.
Also, I am assuming there is a reason for defining it here, is it a best practices thing that Python developers recommend?

File: digi/xbee/xsocket.py [hunk 4]
```diff
@@ -0,0 +1,814 @@
...
+            # Register internal socket state and data reception callbacks.
+            self.__register_state_callback()
+            self.__register_data_received_callback()
+        finally:
+            # Always remove the socket state callback.
+            self.__xbee_device.del_socket_state_received_callback(socket_state_received_callback)
+
+    def bind(self, address):
+        """
+        Binds the socket to address. The socket must not already be bound.
```
Comment (hgonzaleDigi) [inline digi/xbee/xsocket.py]: NP: Binds the socket to the given address

File: digi/xbee/xsocket.py [hunk 5]
```diff
@@ -0,0 +1,814 @@
...
+        # If the socket is not created, create it first.
+        if self.__socket_id is None:
+            self.__create_socket()
+
+        # Create, send and check the socket create packet.
+        bind_packet = SocketBindListenPacket(self.__xbee_device.get_next_frame_id(), self.__socket_id, port)
+        response_packet = self.__xbee_device.send_packet_sync_and_get_response(bind_packet, self.__timeout)
+        self.__check_response(response_packet)
+
+        # Register the data reception from.
```
Comment (hgonzaleDigi) [inline digi/xbee/xsocket.py]: NP: Register the internal data 'reception from' callback

File: digi/xbee/xsocket.py [hunk 6]
```diff
@@ -0,0 +1,814 @@
...
+        # Store the source port.
+        self.__source_port = port
+
+    def listen(self, backlog=1):
+        """
+        Enables a server to accept connections.
+
+        Args:
+            backlog (Integer, optional): The number of unaccepted connections that the system will allow before refusing
+                new connections. If it's lower, it will be set to 0.
```
Comment (hgonzaleDigi) [inline digi/xbee/xsocket.py]: What do you mean by if it's lower? Is it referring to the number of unaccepted connections being lower than the value specified, or vice-versa?

File: digi/xbee/xsocket.py [hunk 7]
```diff
@@ -0,0 +1,814 @@
...
+        received_packet = list()
+
+        # Define the IPv4 client callback.
+        def ipv4_client_callback(packet):
+            if not isinstance(packet, SocketNewIPv4ClientPacket) or packet.socket_id != self.__socket_id:
+                return
+
+            # Add the packet to the list and notify the lock.
+            received_packet.append(packet)
+            lock.acquire()
```
Comment (hgonzaleDigi) [inline digi/xbee/xsocket.py]: These three lines, as well as the ones used  to wait, are repeated several times, wouldn't it be better to create a pair of methods called "notify_lock(lock)" and "wait_lock(lock)"?

File: digi/xbee/xsocket.py [hunk 8]
```diff
@@ -0,0 +1,814 @@
...
+        self.__send(data)
+
+    def sendto(self, data, address):
+        """
+        Sends data to the socket. The socket should not be connected to a remote socket, since the destination socket
+        is specified by ``address``.
+
+        Args:
+            data (Bytearray): the data to send.
+            address (Tuple): the address (host name and port) of the s
```
Comment (hgonzaleDigi) [inline digi/xbee/xsocket.py]: The end of the sentence is missing.

`address (Tuple): the address (host name and port) of the destination socket.`

File: digi/xbee/xsocket.py [hunk 9]
```diff
@@ -0,0 +1,814 @@
...
+        self.__unregister_data_received_callback()
+        self.__unregister_data_received_from_callback()
+
+    def setsocketopt(self, option, value):
+        """
+        Sets the value of the given socket option.
+
+        Args:
+            option (:class:`.SocketOption`): the socket option to set its value.
+            value (Bytearray): the new value of the socket option.
```
Comment (hgonzaleDigi) [inline digi/xbee/xsocket.py]: What values are allowed to be set here? If they are a fixed number of values, maybe we should consider creating an enumeration and then check if the given value matches a valid one, if not, return with an exception.

File: digi/xbee/xsocket.py [hunk 10]
```diff
@@ -0,0 +1,814 @@
...
+
+    def __register_state_callback(self):
+        """
+        Registers the socket state callback to be notified when an error occurs.
+        """
+        if self.__socket_state_callback:
+            return
+
+        def socket_state_callback(socket_id, state):
+            if socket_id != socket_id:
```
Comment (hgonzaleDigi) [inline digi/xbee/xsocket.py]: I guess this is comparing the same variable, hence it is never going to enter through the return.
`if self.__socket_id != socket_id:`

File: digi/xbee/xsocket.py [hunk 11]
```diff
@@ -0,0 +1,814 @@
...
+                return
+            if state != SocketState.CONNECTED:
+                self.__connected = False
+                self.__socket_id = None
+                self.__source_port = None
+                self.__data_received = bytearray()
+                self.__data_received_from_dict = OrderedDict()
+                self.__unregister_state_callback()
```
Comment (hgonzaleDigi) [inline digi/xbee/xsocket.py]: Why are all the callbacks unregistered at this point?

File: digi/xbee/xsocket.py [hunk 12]
```diff
@@ -0,0 +1,814 @@
...
+
+    def __register_data_received_callback(self):
+        """
+        Registers the data received callback to be notified when data is received in the socket.
+        """
+        if self.__data_received_callback:
+            return
+
+        def data_received_callback(socket_id, payload):
+            if socket_id != socket_id:
```
Comment (hgonzaleDigi) [inline digi/xbee/xsocket.py]: Same as above

File: digi/xbee/xsocket.py [hunk 13]
```diff
@@ -0,0 +1,814 @@
...
+    def __register_data_received_from_callback(self):
+        """
+        Registers the data received from callback to be notified when data from a specific address is received
+        in the socket.
+        """
+        if self.__data_received_from_callback:
+            return
+
+        def data_received_from_callback(socket_id, address, payload):
+            if socket_id != socket_id:
```
Comment (hgonzaleDigi) [inline digi/xbee/xsocket.py]: Same as above

File: digi/xbee/xsocket.py (hunk 11 above)
Comment (rubenmoral) [reply digi/xbee/xsocket.py]: Because all other states mean error, and when that happens, the socket is closed.

File: digi/xbee/xsocket.py [hunk 14]
```diff
@@ -0,0 +1,814 @@
...
+        self.__check_response(response_packet)
+
+        # Store the received socket ID.
+        self.__socket_id = response_packet.socket_id
+
+    def __register_state_callback(self):
+        """
+        Registers the socket state callback to be notified when an error occurs.
+        """
+        if self.__socket_state_callback:
```
Comment (rubenmoral) [inline digi/xbee/xsocket.py]: `if self.__socket_state_callback is None:`?

File: examples/communication/socket/SocketUDPServerClientSample/readme.txt [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+
+  Running the example
+  -------------------
+  First, build and launch the application. Then, wait for the example to
+  start the UDP server and send a request to the echo server. The application
+  displays the following output:
+
+      - Starting UDP server at port 4660
+      - Starting UDP server at port 4660
```
Comment (rubenmoral) [inline examples/communication/socket/SocketUDPServerClientSample/readme.txt]: Duplicated line.

**Pull request merged**